### PR TITLE
Do not call RAND_get0_public from within the FIPS provider initialization

### LINKED ISSUE #14437: Intermittent failures due to memleaks detected in threads_test in sanitized builds
The recent addition of concurrent provider loading in the threads_test reveals new intermittent failures in sanitized builds. There are memleaks being detected. 

The problem was most probably already there in the code just the new test case is revealing it.

Shane had this comment:

@mattcaswell or @paulidale see the Sanitizer failure here? (I had this in my recent PR also)
I think test_threads was recently updated - I assume this is finding a race condition now?

This particular line in fipsprov.c indicates that something pretty bad is going on..
```
/* TODO(3.0): Tests will hang if this is removed */
(void)RAND_get0_public(libctx);
```

### PR DESCRIPTION
**Pull request by t8m**
This is necessary to avoid leaks on exit when the FIPS provider is initialized first in other thread than the main one.

Fixes #14437

### CODE REVIEW

**Review by paulidale — APPROVED** [review]
Nice find.

**Comment by mattcaswell** [comment]
Hmmm. It feels wrong that we should need to do this. Does this indicate some underlying bug in the init code?

**Comment by t8m** [comment]
> Hmmm. It feels wrong that we should need to do this. Does this indicate some underlying bug in the init code?

The leaked data are from the initialization of the FIPS provider invoked from the FIPS provider shared library constructor.

The leak does not happen if FIPS provider is not configured in config file.

So yeah, this PR might be papering over a real issue.

The code path is something like this:

1. new thread created
2. it loads default provider
3. this triggers implicit initialization of libcrypto loading the config file
4. this triggers loading the FIPS provider as it is configured to be loaded
5. this runs constructor of the libfips.so
6. this triggers POST routine which initializes a libctx for the FIPS provider and drbg in it

The data from 6 are leaked.

This does not happen if the step 1 is avoided - i.e., the initial load and implicit initialization is done in the main thread.

**Comment by t8m** [comment]
I'll try to look into it some more as I am curious why there is the difference between running this in the main thread and a created thread.

**Comment by richsalz** [comment]
Something shoudl probably be documented, perhaps in `doc/man7/openssl-threads.pod`.  It's not uncommon for multi-threaded programs to require initialization in the main thread.

**Comment by t8m** [comment]
My summary of investigation after many debugging runs:

- The leaked data is just associated with missing deallocation of the public and primary drbgs in the FIPS context.
- It can be reproduced by explicitly loading the FIPS provider as well (if you delete the activate=1 line from fipsmodule.cnf and load "fips" provider by OSSL_PROVIDER_load in a thread).
- More interestingly there is even bigger memory leak triggered if I just run RAND_get0_public() in the thread workers.
- But this memory leak does not happen if I first do OSSL_PROVIDER_load/unload("default") in the thread workers.
- But it happens again if I do RAND_get0_public() without the OSSL_PROVIDER_load/unload("default") in the main thread
- But it does not happen if I do RAND_get0_public() without the OSSL_PROVIDER_load/unload("default") in the main thread and not do any RAND_get0_public() in the thread workers.

Nice...

**Comment by mattcaswell** [comment]
So, do you now understand this problem sufficiently well to fix it?

**Comment by richsalz** [comment]
> So, do you now understand this problem sufficiently well to fix it?

Or, perhaps, what to tell app developers on how to minimize the memory leaks or not worry about them.

**Comment by t8m** [comment]
> So, do you now understand this problem sufficiently well to fix it?

Not yet fully. Unfortunately the leak by RAND_get0_public() without OSSL_PROVIDER_load() is actually a different problem which I fixed in the pushed commit.

Doing the global init is definitely a wrong fix though so I've dropped that.

**Comment by t8m** [comment]
This fixes the leak problem. However I suppose the problem would reappear if we decided that we have to always run KATs. The reason for the leak if the RAND_get0_public is called there was that the prov->provctx on the libcrypto side is not yet initialized when the ossl_init_thread_start is called for the ossl_ctx_thread_stop handler on the libctx from within the provider.

It is possible there is some better way how to fix the underlying problem however this at least is not just papering over the issue as the RAND_get0_public() is really not needed there anymore and it solves the leak.

**Review by paulidale — APPROVED** [review]
Looks okay.

**Comment by slontis** [comment]
The only reason this line was still there was that it was causing errors without it in the tests at some point. It looks like this issue may have shifted now..

**Comment by t8m** [comment]
I'm marking this urgent as this fixes the annoying CI failures. However I'd like to ask @mattcaswell to review before I'll merge.

**Comment by paulidale** [comment]
Urgent is good.

**Comment by mattcaswell** [comment]
I am quite sceptical about the need for these OPENSSL_init_crypto calls in the bio, engine and store code. The `OPENSSL_INIT_BASE_ONLY` option does 3 things:

- Creates the init lock
- Runs OPENSSL_cpuid_setup()
- Initialises the thread system we use for cleaning up thread local resources

bio/engine/store don't care about the init lock. That's entirely internal to the init code. They don't have any assembler code so shouldn't care about cpuid setup, and they don't use any thread local storage.

The previous calls where 0 was passed as the option did precisely nothing at all. So I'd like to know if anything breaks if you remove those calls altogether.

The calls in the err and rand code do make sense because both of those sub-systems utilise thread local storage.

**Comment by t8m** [comment]
OK, I'll removed those as suggested. @paulidale @mattcaswell please re-review.

**Review by mattcaswell — APPROVED** [review]
LGTM. But lets see the CIs go through before merging to make sure the removals haven't had some unexpected side effect.

File: crypto/store/store_init.c [hunk 1]
```diff
@@ -14,7 +14,7 @@
 static CRYPTO_ONCE store_init = CRYPTO_ONCE_STATIC_INIT;
 DEFINE_RUN_ONCE_STATIC(do_store_init)
 {
-    return OPENSSL_init_crypto(0, NULL);
+    return 1;
```
Comment (mattcaswell) [inline crypto/store/store_init.c:17]: Actually.....can we remove this run once altogether now?

Comment (t8m) [reply crypto/store/store_init.c:17]: I thought you'll ask. :grin:

**Review by mattcaswell — APPROVED** [review]
LGTM. Subject to CIs passing.

**Review by paulidale — APPROVED** [review]

**Comment by t8m** [comment]
Aargh, the non-caching acvp test failure is actually valid.

**Comment by t8m** [comment]
Fortunately this is just wrong expectation in acvp_test and is easily fixable.

**Comment by t8m** [comment]
Heh, so re-review needed again. @paulidale @mattcaswell

**Review by mattcaswell — APPROVED** [review]

**Comment by t8m** [comment]
Merged to master. This hopefully fixed the remaining intermittent failures. Or at least the common ones.

**Comment by t8m** [comment]
Thank you for the reviews

**Comment by bernd-edlinger** [comment]
This does actually create a memory leak.
Previously `OPENSSL_init_crypto(OPENSSL_INIT_BASE_ONLY, NULL)` was only called
in `ERR_get_state(void)` and `err_shelve_state(void **state)`.
But ERR_get_state does finally also `OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CRYPTO_STRINGS, NULL);`
which is guranteed to install the `atexit(OPENSSL_cleanup)`, even if there may be a failure in loading the
crypto strings.
err_shelve_state is only called from OPENSSL_init_crypto, and thus the OPENSSL_cleanup handler is also
installed.

But now the OPENSSL_cleanup is no longer called.  Also it is not true that `PENSSL_init_crypto(0, NULL)`
is a no-op, since it was always done at places where global objects are allocated that are only
cleaned up when OPENSSL_cleanup actually runs.

**Pull request closed without merging**